### PR TITLE
Support serializing NativeFontHandle on macOS

### LINKED ISSUE #2051: Implement NativeFontHandle writing on Mac
We can implement this using the technique here: https://searchfox.org/mozilla-central/source/gfx/2d/ScaledFontMac.cpp#255

I have patches in progress.

### PR DESCRIPTION
**Pull request by jrmuizel**
This fixes #2051

---
This change is [Reviewable](https://reviewable.io/reviews/servo/webrender/2122)

### CODE REVIEW

**Comment by glennw** [comment]
Looks like CI failures on Windows at least?

**Comment by jrmuizel** [comment]
I just realized this ends up writing out a new font every time it's used

**Comment by glennw** [comment]
@jrmuizel Is this ready for review now?

**Comment by jrmuizel** [comment]
Yes.

File: wrench/src/cgfont_to_data.rs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+use std;
+use std::io::Cursor;
+use std::io::Read;
+use std::io::Write;
+
+
+fn calc_table_checksum<D: Read>(mut data: D) -> u32 {
+    let mut sum: u32 = 0;
+    loop {
+        if let Ok(x) = data.read_u32::<BigEndian>() {
```
Comment (kvark) [inline wrench/src/cgfont_to_data.rs]: `while let Ok(x) = ... {...}`

File: wrench/src/cgfont_to_data.rs [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+            data.read(&mut buf).unwrap();
+            sum = sum.wrapping_add(BigEndian::read_u32(&buf));
+            break;
+        }
+    }
+    sum
+}
+
+fn max_pow_2_less_than(a: u16) -> u16
+{
```
Comment (kvark) [inline wrench/src/cgfont_to_data.rs]: nit: opening bracket

File: wrench/src/cgfont_to_data.rs [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    let x = 1;
+    let mut shift = 0;
+    while (x << (shift + 1)) < a {
+        shift += 1;
+    }
+    return shift;
```
Comment (kvark) [inline wrench/src/cgfont_to_data.rs]: nit: return

File: wrench/src/cgfont_to_data.rs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    let mut cff = false;
+    let tags = font.copy_table_tags();
+    let count = tags.len() as u16;
+
+    let mut records = Vec::new();
+    let mut offset: u32 = 0;
+    offset += 4 * 3;
+    offset += 4 * 4 * (count as u32);
+    for tag in tags.iter() {
+        if tag == 0x43464620 { // 'CFF '
```
Comment (kvark) [inline wrench/src/cgfont_to_data.rs]: nit: could be `const` thing for clarity

File: wrench/src/yaml_frame_writer.rs [hunk 5]
```diff
@@ -217,25 +217,49 @@ fn write_sc(parent: &mut Table, sc: &StackingContext, properties: &SceneProperti
 }
 
 #[cfg(target_os = "windows")]
-fn native_font_handle_to_yaml(handle: &NativeFontHandle, parent: &mut yaml_rust::yaml::Hash) {
+fn native_font_handle_to_yaml(_rsrc: &mut ResourceGenerator,
+                              handle: &NativeFontHandle,
```
Comment (kvark) [inline wrench/src/yaml_frame_writer.rs]: nit: formatting

File: wrench/src/yaml_frame_writer.rs [hunk 6]
```diff
@@ -217,25 +217,49 @@ fn write_sc(parent: &mut Table, sc: &StackingContext, properties: &SceneProperti
...
 }
 
 #[cfg(target_os = "macos")]
-fn native_font_handle_to_yaml(_: &NativeFontHandle, _: &mut yaml_rust::yaml::Hash) {
-    panic!("Can't native_handle_to_yaml on this platform");
+fn native_font_handle_to_yaml(rsrc: &mut ResourceGenerator,
+                              handle: &NativeFontHandle,
+                              parent: &mut yaml_rust::yaml::Hash,
+                              path_opt: &mut Option<PathBuf>) {
+    let path = if let &mut Some(ref path) = path_opt {
```
Comment (kvark) [inline wrench/src/yaml_frame_writer.rs]: nit: use `match *path_opt`

File: wrench/src/yaml_frame_writer.rs [hunk 7]
```diff
@@ -254,11 +278,30 @@ struct CachedImage {
...
 }
 
+struct ResourceGenerator {
+    base: PathBuf,
+    next_num: u32,
+    prefix: String,
+}
+
+impl ResourceGenerator {
+    fn next_rsrc_paths(&mut self, base: &str, ext: &str, ) -> (PathBuf, PathBuf) {
```
Comment (kvark) [inline wrench/src/yaml_frame_writer.rs]: nit: trailing comma?

**Comment by glennw** [comment]
I'm just going to assume that the code in ```cgfont_to_data``` makes sense. I guess it's based on some existing code? If so, it's probably worth adding a comment in the code referencing where it's ported from.

@bors-servo r+

**Pull request merged**